### PR TITLE
Discover, tar, and deploy apps

### PR DESCRIPTION
**Pull request by leeweisberger**
## Description
Discover apps (behind a flag).

Once discovered, tar and upload app content (same as we do for tasks).

Add discovered apps to a deployment along with the url of the uploaded content.

## Test plan
Tested locally. Some unit tests added.

### CODE REVIEW

File: pkg/cmd/tasks/deploy/deployer.go [hunk 1]
```diff
@@ -120,6 +121,41 @@ func (d *deployer) DeployTasks(ctx context.Context, taskConfigs []discover.TaskC
...
+		// Get the root directory of the git repo with which the task is associated.
+		var gitRoot string
+		if repo != nil {
+			w, err := repo.Worktree()
+			if err == nil {
+				gitRoot = w.Filesystem.Root()
+			}
+		}
+		gitRoots[gitRoot] = true
+	}
```
Comment (leeweisberger) [inline pkg/cmd/tasks/deploy/deployer.go:158]: This is generally copied from tasks code

File: pkg/cmd/tasks/deploy/deployer.go [hunk 2]
```diff
@@ -120,6 +121,41 @@ func (d *deployer) DeployTasks(ctx context.Context, taskConfigs []discover.TaskC
...
+		}
+
+		appToDeploy := api.DeployApp{
+			ID:          ac.ID,
+			UploadID:    uploadIDs[ac.ID],
+			GitFilePath: filePath,
+		}
+		if err != nil {
+			return err
+		}
```
Comment (stepzhou) [inline pkg/cmd/tasks/deploy/deployer.go:146]: Is this doing anything?

File: pkg/cmd/tasks/deploy/deployer.go [hunk 3]
```diff
@@ -120,6 +121,41 @@ func (d *deployer) DeployTasks(ctx context.Context, taskConfigs []discover.TaskC
...
+		if repo != nil {
+			w, err := repo.Worktree()
+			if err == nil {
+				gitRoot = w.Filesystem.Root()
+			}
+		}
+		gitRoots[gitRoot] = true
+	}
 
 	// If tasks in a single deploy come from different git repos, we do not
```
Comment (stepzhou) [inline pkg/cmd/tasks/deploy/deployer.go:160]: tasks or apps

File: pkg/cmd/tasks/deploy/deployer.go [hunk 4]
```diff
@@ -120,6 +121,41 @@ func (d *deployer) DeployTasks(ctx context.Context, taskConfigs []discover.TaskC
 		}
 		gitRoots[gitRoot] = true
 	}
+	var appsToDeploy []api.DeployApp
+	for _, ac := range appConfigs {
+		repo, err = d.repoGetter.GetGitRepo(ac.Entrypoint)
```
Comment (stepzhou) [inline pkg/cmd/tasks/deploy/deployer.go:126]: I'm confused what this Entrypoint is. I thought the app entrypoints were that gcs:// string. Is this entrypoint different?

**Review by stepzhou — CHANGES_REQUESTED** [review]
Requesting for my entrypoint question

File: pkg/cmd/tasks/deploy/deployer.go (hunk 4 above)
Comment (stepzhou) [reply pkg/cmd/tasks/deploy/deployer.go:126]: Oh I get it. This entrypoint is the path that's defined in the app definition yaml file. The terminology here is really confusing

**Review by stepzhou — APPROVED** [review]

File: pkg/cmd/tasks/deploy/deployer.go (hunk 4 above)
Comment (leeweisberger) [reply pkg/cmd/tasks/deploy/deployer.go]: I just removed entrypoint. I have no clue whether it's going to be useful. We can decide later.

**Pull request merged**